Add some tests about proactive cancel

sbernard31 opened the pull request:
I added some tests relative to #688.

I think "a test about pro active cancel (ed40c69)" makes sense to be added.
The second one, "test that response with observe option is for simple GET (8fcf2cb)", maybe a little less, but I used it to check #689.

boaks commented on californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java:
```diff
@@ -1327,4 +1327,52 @@ public void testNotificationIsNotHandledAsProactiveCancelResponse() throws Excep
 	}
+
+	/**
+	 * Ensure that response with unexpected observe option is not ignored
```
Can you add the hint, that this is only intended to ensure compatibility with a broken/none RFC-7252 compliant client?

sbernard31 replied:
I agree this test is a bit strange and a bit out of spec, I mainly write it to ensure that the issue reported in #688 was fixed.

I can understand, if you think this should not be integrated in the californium test base.

boaks replied:
From my side there is no objection to have this test included to ensure, your already analysed issue is still fixed. A added hint would ensure, that no one takes this as compliant example.

sbernard31 replied:
Ok I will add this !

boaks replied:
> Ok I will add this !

Any due day? I would prefer to have such a comment added before merging

boaks commented on californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java:
```diff
@@ -1256,4 +1256,75 @@ public void testObserveFailureBeforeToSendDuringBlockNotification() throws Excep
+	 * </pre>
+	 */
+	@Test
+	public void testNotificationIsNotHandledAsProactiveCancelResponse() throws Exception {
```
I would prefer to have different tests for:
- receiving notifications during the cancel request.
- receiving notifications after successful cancel request 
- receiving notifications after failed cancel request

sbernard31 replied:
This test is only about : 
 - receiving notifications during the cancel request.

(I changed it a little bit after our discussion in #688, before to push this PR)

Would you like I add tests for : 
- receiving notifications after successful cancel request
- receiving notifications after failed cancel request

Currently the behavior for failed and successful cancel will be the same.

boaks replied:
> Currently the behavior for failed and successful cancel will be the same.

Do you mean, currently 
- receiving notifications after successful cancel request
- receiving notifications after failed cancel request

behaves equal? 
I would guess (hope :-) ), the client will RST notifications, if the cancel is successful.

sbernard31 replied:
> I would guess (hope :-) ), the client will RST notifications, if the cancel is successful.

After cancelling, notifications are not RST anymore. (success or not)
That's what I try to explain in https://github.com/eclipse/californium/issues/688#issuecomment-403501405.

There is 2 API, coapclient and coapendpoint.
Coapclient is more high level api than coapendpoint.

With coapclient, you can use CoapObserveRelation.proactiveCancel() and it will behave like you expect.

But coapendpoint, is low-level api, there is one API to send proactive cancel request and one to remove observation at client side (passive cancellation).

FMPOV in a perfect world, with coapendpoint API notification should be always handle as long as we don't remove observation from obervation store.
But there is a side effect and during an ongoing proactive cancel request, notifications are RST. 
As you explain fixing this means adding corner case in the stack and maybe this is not really worthy. In generally, the proactive use case is to do both at the same time. I mean removing observation from the store, then sending a cancel request.

With the low level API, the success of cancel request does not determine if the observation is removed or not. This should be done manually before or after using endpoint.cancelObservation()

In this test, I just send the cancel request.

boaks replied:
Sorry Simon,
your tests seems for me to be smoke tests, where messages are fired in a chaotic, none-compliant way. I thought, your intention is to ensure, that there are no leaks introduced. If I have over seen the variant, where the cancel is responded successful and further notifications are accepted instead of rejected, then this is my fault. But even in the comment you refer, the cancel request is NOT responded with success, instead it timesout, which I would count as "failed"

That's also my comment:
Please provide not only the "smoke/chaos" test, provide also the three compliant test-cases.

boaks replied:
After removing "the smoke", I recognized, that this test doesn't use the `CoapClient`. Therefore the "proactive cancel" doesn't work. I prepared a fix for proactive cancel without `CoapClient`, but using "observe" without a `CoapClient` will not filter the notifications based on their order and requires a `NotificationListener`. 
So at least, this seems for me to be not recommended.
I got also aware, that `Request.waitForResponse()` is still in between. The comments about the notifications are not cleaned up and the implementation sticks also on the old notification processing.

sbernard31 replied:
> your intention is to ensure, that there are no leaks introduced.

My intention was to test the use case you described at https://github.com/eclipse/californium/issues/688#issuecomment-402757040

> But even in the comment you refer, the cancel request is NOT responded with success,

I would like to fix this comment but I don't know which one you are talking about. Could you precise which one ?

> Please provide not only the "smoke/chaos" test, provide also the three compliant test-cases.

I think this is not so chaotic, just 2 notifications which comes before we receive the proactive cancel and a lost ACK for the cancel separated response. (This is compliant case)

but I could add more tests, but currently I'm not sure what is the behavior expected.

> After removing "the smoke"

I still don't understand what you consider as smoke in this test. :-/

> but using "observe" without a CoapClient will not filter the notifications based on their order and requires a NotificationListener.
So at least, this seems for me to be not recommended.

I think this is the interest of having this low-level API. Bringing more flexibility.

> I got also aware, that Request.waitForResponse() is still in between. The comments about the notifications are not cleaned up and the implementation sticks also on the old notification processing.

I do not understand the issue here with the use of waitForResponse, I do not want to get notification, I want to ensure that notification received is not consider as response for the proactive cancel request.

boaks replied:
"chaotic" means, it's not that easy/obvious, what it tests. At least for me ...

sbernard31 replied:
> "chaotic" means, it's not that easy/obvious, what it tests. At least for me ...

I do not take your remarks badly, I just don't know how to do this better. :-)
How could I test that a notification is not handled as a proactive cancel response in a more easy/obvious to understand way ?

(Maybe just remove the lost ACK ?)

boaks replied:
> I would like to fix this comment but I don't know which one you are talking about. Could you precise which one ?

```
X------ ACK M3              ------ // proactive ACK lost
<------ CON M6, T1          ------ // response of proactive cancel
```

That's what is "chaotic" for me. First "ACK lost", and then response. So the test-case doesn't simply piggybacked response to the request. That makes it hard to see, if it fails, because the test is broken or because the implementation is broken. To understand such stuff, much more time is required.

sbernard31 replied:
OK I changed this.

boaks replied:
> OK I changed this.

Any due day? For that issue, we may also merge it (after https://github.com/eclipse/californium/pull/692#discussion_r201326176) and add this stuff later.

sbernard31 replied:
I would like to finished some tasks, I postponed too many time for Leshan. I hope I could rework on this before the end of the day.

Did you ask this because you would like to release the M11 ? If yes, I think the tests in this PR is not mandatory for M11.

boaks replied:
Mentally, I shifted M11 already for a week :-).

boaks commented:
After changing the a little the threading (see issue #690), I struggling with some race-conditions.
Would it be possible, to add `@RunWith(RepeatingTestRunner.class)` locally and check the tests for robustness? I don't say, there are issues with that, but I would prefer, if it get's checked.

sbernard31 commented:
About @RunWith(RepeatingTestRunner.class), default value make the tests very long. Do you advice to keep this default value or to reduce it ?

sbernard31 commented:
I added comment on f605ca0  to explain this is a pretty buggy use case.

About proactive cancel tests: 
- I added a "standard" test about proactive cancel.
- for all tests, I `cancelObservation` before as it seems to be the right/standard way to do. 
- I also tried to make tests less smoky.

For the last commit, I think we should cancelObservation before to send proactive request in coapclient. If you think, I'm wrong I will just not integrate it. (not a big deal)

I ran the 3 tests with  @RunWith(RepeatingTestRunner.class) in eclipse. I don't know if this really what you expect. It sounds good.

boaks commented:
> make the tests very long

Sure, it repeats the tests 100 times until it fails :-). 
The idea is, that at the end, before merging, you add locally the `@RunWith(RepeatingTestRunner.class)` and do a final check with the Eclipse Test Runner only running that test class. If that passes, don't commit this last change (otherwise the eclipse jenkins will be in deep trouble :-) ).
    
Just to mention.
My race-conditions where caused by 2 protocol threads, I fixed that ensuring the usage of 1 protocol thread again.

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java:
```diff
@@ -203,9 +203,8 @@ private void cancel() {
-		sendCancelObserve();
-		// cancel observe relation
 		cancel();
+		sendCancelObserve();
```
The other order was chosen, to ensure, that the token is not used in the "meantime" (between line 206 and 207) by an other request. But it's academically and not really analysed, if this is really a risk.

sbernard31 replied:
I feel you're right that we may have a potential token collision here, but I think that's was already the case before because `sendCancelObserve` is not synchronous.

If we want to avoid this, the observationStore should contains the observation (and so lock the token) until we complete the proactive cancel exchange.

If we want to do that we should go with something more like you do in #694.

So, I will not integrate this last commit and merge only the tests.

boaks commented:
> I ran the 3 tests with @RunWith(RepeatingTestRunner.class) in eclipse. I don't know if this really what you expect. It sounds good.

Cool, does it mean you run the 3 tests each 100 times successful local on your PC?
Yes, that's what I think is a good practice, at least for new stuff.
 
LGTM

sbernard31 commented:
> Cool, does it mean you run the 3 tests each 100 times successful local on your PC?

Yep.

sbernard31 commented:
f605ca043d1bcafe0f8809a9e7baf8619054da79 and 06727905d8b6cb02d6e98e51533f42cedc238459 are integrated in **2.0.x**.

So I close this PR.
